Display license out of compliance warnings

bernardjkim opened the pull request:
This PR updates `tsh` to display license out of compliance warnings on login. For now the only possible out of compliance warning is an expired license. Contributes to https://github.com/gravitational/cloud/issues/1673.

## Testing
```sh
$ tsh login --proxy tele.example.com --user teleport-admin
Enter password for Teleport user teleport-admin:
Enter your OTP token:
> Profile URL:        https://tele.example.com:443
  Logged in as:       teleport-admin
  Cluster:            tele.example.com
  Roles:              access, editor
  Logins:             root, -teleport-internal-join
  Kubernetes:         enabled
  Valid until:        2022-07-22 02:17:40 -0700 PDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

Your Teleport license has expired. If you are the System Administrator, please reach out to your Account Manager and obtain a new license to continue using Teleport.
```

bernardjkim committed:
Display license out of compliance warnings

zmb3 commented:
I'm not sure I understand. Why do we need to write to disk for this?

zmb3 commented:
Do you think there's some way the server can send this info back in the existing API calls so the client doesn't have to make a separate call for license status?

If I were using tsh and my administrators failed to update the license in time I would be annoyed that tsh is leaving behind these random files in my home directory.

bernardjkim commented:
> If I were using tsh and my administrators failed to update the license in time I would be annoyed that tsh is leaving behind these random files in my home directory.

I agree this would be annoying, but the tsh home directory is at least cleaned up on logout.

> Do you think there's some way the server can send this info back in the existing API calls so the client doesn't have to make a separate call for license status?

I'll take another look. I'm still not too familiar with the tsh/tctl code.

jimbishopp commented:
@bernardjkim Is there a requirement to display the warning for every command?

bernardjkim commented:
> @bernardjkim Is there a requirement to display the warning for every command?

I was following the specifications [here](https://github.com/gravitational/webapps.e/issues/314#issue-1269623941). If it doesn't make sense to display the warnings for all commands, we could open up a discussion about it.

zmb3 commented:
I don't think you'd want to add it to the `SSHLoginResponse`, the GRPC API seems more appropriate.

I'm wondering if you can return an "error" along with valid data, or use some sort of GRPC interceptor to augment the response with extra metadata.

bernardjkim committed:
No longer write license status to disk

tsh/tctl now read Ping RPC metadata for license compliance warnings

bernardjkim committed:
Update e repo

bernardjkim committed:
Merge branch 'master' into bernard/license-warnings

bernardjkim commented:
Hey @zmb3 how does [6b1be0f](https://github.com/gravitational/teleport/pull/14114/commits/6b1be0fb8c10b4001271a7adb28bfcbbfd49674d) look?

Teleport Enterprise now appends license warning messages to every `Ping` rpc response https://github.com/gravitational/teleport.e/pull/490. `tsh/tctl` now make a `Ping` request on every command. The license status is no longer written to disk.

mcbattirola reviewed (commented):
We mentioned in the UX weekly something about only displaying the license warning on `tsh login`, IIRC @xinding33 said this is acceptable. That way we wouldn't have to ping the server on every command, how does that sound?

zmb3 commented:
Yeah, I think for `tsh` we're happy with on login only (this is a typically a daily event).

Not sure about `tctl` yet.

bernardjkim committed:
Display license warnings only on login

bernardjkim committed:
Fix lint warning

bernardjkim committed:
Bump e

bernardjkim committed:
Merge branch 'master' into bernard/license-warnings

bernardjkim commented:
> We mentioned in the UX weekly something about only displaying the license warning on tsh login

Displaying the warnings only on `tsh` login makes sense to me. I think we can go ahead and also drop the warnings for `tctl` then. `tctl` is only used for admin purposes. Users will need to use `tsh` or the webUI to interact with their Teleport resources.

bernardjkim committed:
Chain plugin interceptors first

- Plugin interceptor should be outer most wrapper around the rpc.
- Plugin interceptor does not need to register more than one interceptor
  for now.

bernardjkim committed:
Merge branch 'master' into bernard/license-warnings

bernardjkim committed:
Bump e

bernardjkim committed:
Bump e

bernardjkim committed:
Capitalize logs

bernardjkim commented:
@zmb3 

Sorry for taking so long with this. Unit tests should be passing now. Any concerns with the implementation of the graph interceptors?

zmb3 commented:
No worries, @bernardjkim. 

If we're putting license warnings directly in the ping response, do we even need to use a GRPC interceptor or can we just add some logic to the ping RPC?

bernardjkim committed:
Revert interceptor implementation

bernardjkim committed:
Embed enforcer service into auth server

- PingResponse message now includes a LicenseWarnings field.
- The Ping rpc now pulls license status from the embedded enforcer
  service and includes license warnings in the ping response.

bernardjkim commented:
We don't need to use an interceptor. Alternatively, we can embed an optional enforcer service into the auth server. Then we can modify the ping rpc to pull any license warnings from the enforcer into the ping response. How does [8a16b73](https://github.com/gravitational/teleport/pull/14114/commits/8a16b73375d4be0b652790317b816fc1f7a5f1ce) look?

bernardjkim commented:
Bump @zmb3

zmb3 commented on lib/services/local/enforcer.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/gravitational/reporting/types"
+)
+
+// NoopEnforcerService is a no-op enforcer service.
```
```suggestion
// NoopEnforcer is a no-op enforcer.
```

The interface is called `Enforcer`, so let's just call this `NoopEnforcer`.

zmb3 commented on lib/auth/auth_with_roles.go:
```diff
@@ -1920,13 +1920,23 @@ func (a *ServerWithRoles) Ping(ctx context.Context) (proto.PingResponse, error)
+	if err != nil {
+		return proto.PingResponse{}, trace.Wrap(err)
+	}
+	warnings := make([]string, 0)
```
```suggestion
	var warnings []string
```

Make with an explicit len makes sense if you know ahead of time the exact length you want, but you don't. So I'd stick with this instead.

zmb3 commented on api/constants/constants.go:
```diff
@@ -134,6 +134,9 @@ const (
 	SSHRSAType = "ssh-rsa"
+
+	// LicenseExpiredType is defines a license expired notification
+	LicenseExpiredType = "licenseExpired"
```
```suggestion
	LicenseExpiredNotification = "licenseExpired"
```

I'm not a fan of using this `constants` package unless you have something that really needs to be accessed in many places throughout the app.

Since licensing is a function of the auth server only, putting this in `lib/auth` or maybe even `api` might make more sense. Up to you.

bernardjkim replied:
That makes sense. For now we only need to access the license notifications in the `/lib/auth` package so we can localize it there.

zmb3 reviewed (approved):
Looks better, thanks for your patience on this one!

See comments for some minor suggestions.

jimbishopp reviewed (approved):
Looks good. Minor comments.

bernardjkim commented on tool/tsh/tsh.go:
```diff
@@ -1448,7 +1448,27 @@ func onLogin(cf *CLIConf) error {
+		return trace.Wrap(err)
+	}
+
+	if err := displayLicenseWarnings(cf.Context, tc); err != nil {
+		log.WithError(err).Debug("Failed to display license warnings.")
+	}
+
+	return nil
+}
+
+// displayLicenseWarnings displays license out of compliance warnings.
+func displayLicenseWarnings(ctx context.Context, tc *client.TeleportClient) error {
+	resp, err := tc.Ping(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, warning := range resp.LicenseWarnings {
+		fmt.Fprintln(os.Stderr, warning)
+	}
+	return nil
 }
```
The warning message is not delimited by a new line https://github.com/gravitational/houston/blob/bddea97761a32679c9a3bec358da3f9f3698b5a6/lib/web/apiserver.go#L373-L383

I'll add an extra newline so that there is always space between two warning messages. 

I implemented it this way to allow for multiple warning messages, but I'm thinking we should try to avoid having more than one warning message displayed this way. If we do want to display more notifications this way, we should probably put some more consideration into how it should look.

bernardjkim committed:
Address PR feedback

bernardjkim committed:
Merge branch 'master' into bernard/license-warnings

bernardjkim committed:
Bump e

bernardjkim committed:
Merge branch 'master' into bernard/license-warnings

bernardjkim commented:
Thanks for all the feedback!
